File: Arrays/min_jumps.cpp
// https://www.geeksforgeeks.org/problems/minimum-number-of-jumps-1587115620/1
// TC : O(N^2)
// SC : O(N)
// wrong approach right now 
class Solution {
  public:
    int minJumps(vector<int>& arr) {
        int dis=arr.size();
        int i=0;
        int count=0;
        while(dis>0){
            cout << arr[i] << endl;
            count++;
            if(arr[i]==0){
                return -1;
            }
            else{
                int maxi=arr[arr[i]+1];
                int new_i = arr[i];
                for(int j=1;j<=arr[i];i++){
                    if(maxi < arr[j+arr[i]]){
                        maxi = arr[j+arr[i]];
                        new_i = j+arr[i];
                    }
                }
                i=new_i;
                dis-=maxi;
            }
            
            if(dis<=0){
                return count;
            }
        }
    }
};

File: Arrays/min_jumps.java
// https://www.geeksforgeeks.org/problems/minimum-number-of-jumps-1587115620/1
// Time Complexity : O(N)
// Space Complexity: O(1)

class Solution {
    public int minJumps(int[] arr) {
        if (arr.length <= 1)
            return 0;
        int jumps = 0;
        int currentReach = 0;       // End of the range reachable using the current number of jumps.
        int farthestReach = 0;      // Farthest index we can reach while scanning the current range.
        for (int i = 0; i < arr.length - 1; i++) {
            farthestReach = Math.max(farthestReach, i + arr[i]);   // Expand the farthest reachable index.
            if (i == currentReach) {                               // Finished scanning the current reachable range.
                if (farthestReach == currentReach)                 // Couldn't extend the range → stuck.
                    return -1;
                currentReach = farthestReach;                      // Begin exploring the next reachable range.
                jumps++;                                           // One jump is now committed.
            }
        }
        return jumps;
    }
}
